src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -1269,5 +1226,270 @@ internal override bool CanAddProjectReference(ProjectId referencingProject, Proj
+        /// <remarks>This is needed to synchronize with <see cref="ApplyChangeToWorkspace(Action{Workspace})" /> to avoid any races. This
+        /// method could be moved down to the core Workspace layer and then could use the synchronization lock there.</remarks>
+        /// <param name="projectId">The <see cref="ProjectId" /> to change.</param>
+        /// <param name="mutation">A function that, given the old <see cref="CodeAnalysis.Project"/> will produce a new one.</param>
+        public void ApplyBatchChangeToProject(ProjectId projectId, Func<CodeAnalysis.Solution, CodeAnalysis.Solution> mutation)
+        {
+            lock (_gate)
+            {
+                var oldSolution = this.CurrentSolution;
+                var newSolution = mutation(oldSolution);
+
+                if (oldSolution == newSolution)
+                {
+                    return;
+                }
+
+                SetCurrentSolution(newSolution);
+                RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectChanged, oldSolution, newSolution, projectId);
+            }
+        }
+
+        private Dictionary<ProjectId, ProjectReferenceInformation> _projectReferenceInfo = new Dictionary<ProjectId, ProjectReferenceInformation>();
+
+        private ProjectReferenceInformation GetReferenceInfo_NoLock(ProjectId projectId)
+        {
+            return _projectReferenceInfo.GetOrAdd(projectId, _ => new ProjectReferenceInformation());
+        }
+
+        protected internal override void OnProjectRemoved(ProjectId projectId)
+        {
+            lock (_gate)
+            {
+                var project = this.CurrentSolution.GetProject(projectId);
+
+                _projectReferenceInfo.Remove(projectId);
+                _projectToGuidMap = _projectToGuidMap.Remove(projectId);
+                _projectToHierarchyMap = _projectToHierarchyMap.Remove(projectId);
+                _projectNameToProjectMap = _projectNameToProjectMap.Remove(project.Name);
+
+                base.OnProjectRemoved(projectId);
+            }
+        }
+
+        private class ProjectReferenceInformation
+        {
+            public List<string> OutputPaths = new List<string>();
+            public List<(string path, ProjectReference projectReference)> ConvertedProjectReferences = new List<(string path, ProjectReference)>();
+        }
+
+        /// <summary>
+        /// A multimap from an output path to the project outputting to it. Ideally, this shouldn't ever
+        /// actually be a true multimap, since we shouldn't have two projects outputting to the same path, but
+        /// any bug by a project adding the wrong output path means we could end up with some duplication.
+        /// In that case, we'll temporarily have two until (hopefully) somebody removes it.
+        /// </summary>
+        private readonly Dictionary<string, List<ProjectId>> _projectsByOutputPath = new Dictionary<string, List<ProjectId>>(StringComparer.OrdinalIgnoreCase);
```

Comment:
>private readonly Dictionary<string, List<ProjectId>> _projectsByOutputPat [](start = 8, length = 73)

Move to the top.

Reply:
Right now I'm keeping the sections together, and will move things around further once this lands.